// sprotty/lib/features/command-palette/action-providers.js
"use strict";
/********************************************************************************
 * Copyright (c) 2019 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RevealNamedElementActionProvider = exports.CommandPaletteActionProviderRegistry = void 0;
const inversify_1 = require("inversify");
const actions_1 = require("sprotty-protocol/lib/actions");
const action_1 = require("../../base/actions/action");
const types_1 = require("../../base/types");
const iterable_1 = require("../../utils/iterable");
const model_1 = require("../nameable/model");
let CommandPaletteActionProviderRegistry = class CommandPaletteActionProviderRegistry {
    constructor(actionProviders = []) {
        this.actionProviders = actionProviders;
    }
    getActions(root, text, lastMousePosition, index) {
        const actionLists = this.actionProviders.map(provider => provider.getActions(root, text, lastMousePosition, index));
        return Promise.all(actionLists).then(p => p.reduce((acc, promise) => promise !== undefined ? acc.concat(promise) : acc));
    }
};
exports.CommandPaletteActionProviderRegistry = CommandPaletteActionProviderRegistry;
exports.CommandPaletteActionProviderRegistry = CommandPaletteActionProviderRegistry = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.multiInject)(types_1.TYPES.ICommandPaletteActionProvider)),
    __param(0, (0, inversify_1.optional)()),
    __metadata("design:paramtypes", [Array])
], CommandPaletteActionProviderRegistry);
let RevealNamedElementActionProvider = class RevealNamedElementActionProvider {
    constructor(logger) {
        this.logger = logger;
    }
    getActions(root, text, lastMousePosition, index) {
        if (index !== undefined && index % 2 === 0)
            return Promise.resolve(this.createSelectActions(root));
        else
            return Promise.resolve([new action_1.LabeledAction("Select all", [actions_1.SelectAllAction.create()])]);
    }
    createSelectActions(modelRoot) {
        const nameables = (0, iterable_1.toArray)(modelRoot.index.all().filter(element => (0, model_1.isNameable)(element)));
        return nameables.map(nameable => new action_1.LabeledAction(`Reveal ${(0, model_1.name)(nameable)}`, [actions_1.SelectAction.create({ selectedElementsIDs: [nameable.id] }), actions_1.CenterAction.create([nameable.id])], 'eye'));
    }
};
exports.RevealNamedElementActionProvider = RevealNamedElementActionProvider;
exports.RevealNamedElementActionProvider = RevealNamedElementActionProvider = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(types_1.TYPES.ILogger)),
    __metadata("design:paramtypes", [Object])
], RevealNamedElementActionProvider);
