// sprotty/lib/base/views/key-tool.js
"use strict";
/********************************************************************************
 * Copyright (c) 2017-2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyListener = exports.KeyTool = void 0;
const inversify_1 = require("inversify");
const types_1 = require("../types");
const smodel_1 = require("../model/smodel");
const vnode_utils_1 = require("./vnode-utils");
let KeyTool = class KeyTool {
    constructor(keyListeners = []) {
        this.keyListeners = keyListeners;
    }
    register(keyListener) {
        this.keyListeners.push(keyListener);
    }
    deregister(keyListener) {
        const index = this.keyListeners.indexOf(keyListener);
        if (index >= 0)
            this.keyListeners.splice(index, 1);
    }
    handleEvent(methodName, model, event) {
        const actions = this.keyListeners
            .map(listener => listener[methodName].apply(listener, [model, event]))
            .reduce((a, b) => a.concat(b));
        if (actions.length > 0) {
            event.preventDefault();
            this.actionDispatcher.dispatchAll(actions);
        }
    }
    keyDown(element, event) {
        this.handleEvent('keyDown', element, event);
    }
    keyUp(element, event) {
        this.handleEvent('keyUp', element, event);
    }
    focus() { }
    decorate(vnode, element) {
        if (element instanceof smodel_1.SModelRootImpl) {
            (0, vnode_utils_1.on)(vnode, 'focus', this.focus.bind(this, element));
            (0, vnode_utils_1.on)(vnode, 'keydown', this.keyDown.bind(this, element));
            (0, vnode_utils_1.on)(vnode, 'keyup', this.keyUp.bind(this, element));
        }
        return vnode;
    }
    postUpdate() {
    }
};
exports.KeyTool = KeyTool;
__decorate([
    (0, inversify_1.inject)(types_1.TYPES.IActionDispatcher),
    __metadata("design:type", Object)
], KeyTool.prototype, "actionDispatcher", void 0);
exports.KeyTool = KeyTool = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.multiInject)(types_1.TYPES.KeyListener)),
    __param(0, (0, inversify_1.optional)()),
    __metadata("design:paramtypes", [Array])
], KeyTool);
let KeyListener = class KeyListener {
    keyDown(element, event) {
        return [];
    }
    keyUp(element, event) {
        return [];
    }
};
exports.KeyListener = KeyListener;
exports.KeyListener = KeyListener = __decorate([
    (0, inversify_1.injectable)()
], KeyListener);
