// sprotty/lib/features/bounds/bounds-manipulation.js
"use strict";
/********************************************************************************
 * Copyright (c) 2017-2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestBoundsCommand = exports.SetBoundsCommand = void 0;
const inversify_1 = require("inversify");
const actions_1 = require("sprotty-protocol/lib/actions");
const command_1 = require("../../base/commands/command");
const types_1 = require("../../base/types");
const model_1 = require("./model");
let SetBoundsCommand = class SetBoundsCommand extends command_1.SystemCommand {
    constructor(action) {
        super();
        this.action = action;
        this.bounds = [];
    }
    execute(context) {
        this.action.bounds.forEach(b => {
            const element = context.root.index.getById(b.elementId);
            if (element && (0, model_1.isBoundsAware)(element)) {
                this.bounds.push({
                    element: element,
                    oldBounds: element.bounds,
                    newPosition: b.newPosition,
                    newSize: b.newSize
                });
            }
        });
        return this.redo(context);
    }
    undo(context) {
        this.bounds.forEach(b => b.element.bounds = b.oldBounds);
        return context.root;
    }
    redo(context) {
        this.bounds.forEach(b => {
            if (b.newPosition)
                b.element.bounds = Object.assign(Object.assign({}, b.newPosition), b.newSize);
            else
                // keep the position
                b.element.bounds = Object.assign({ x: b.element.bounds.x, y: b.element.bounds.y }, b.newSize);
        });
        return context.root;
    }
};
exports.SetBoundsCommand = SetBoundsCommand;
SetBoundsCommand.KIND = actions_1.SetBoundsAction.KIND;
exports.SetBoundsCommand = SetBoundsCommand = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(types_1.TYPES.Action)),
    __metadata("design:paramtypes", [Object])
], SetBoundsCommand);
let RequestBoundsCommand = class RequestBoundsCommand extends command_1.HiddenCommand {
    constructor(action) {
        super();
        this.action = action;
    }
    execute(context) {
        return {
            model: context.modelFactory.createRoot(this.action.newRoot),
            modelChanged: true,
            cause: this.action
        };
    }
    get blockUntil() {
        return action => action.kind === actions_1.ComputedBoundsAction.KIND;
    }
};
exports.RequestBoundsCommand = RequestBoundsCommand;
RequestBoundsCommand.KIND = actions_1.RequestBoundsAction.KIND;
exports.RequestBoundsCommand = RequestBoundsCommand = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(types_1.TYPES.Action)),
    __metadata("design:paramtypes", [Object])
], RequestBoundsCommand);
