// sprotty/lib/index.js
"use strict";
/********************************************************************************
 * Copyright (c) 2017-2024 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.modelSourceModule = exports.zorderModule = exports.viewportModule = exports.updateModule = exports.undoRedoModule = exports.selectModule = exports.routingModule = exports.openModule = exports.moveModule = exports.hoverModule = exports.fadeModule = exports.exportModule = exports.expandModule = exports.edgeLayoutModule = exports.edgeJunctionModule = exports.edgeIntersectionModule = exports.decorationModule = exports.contextMenuModule = exports.commandPaletteModule = exports.buttonModule = exports.boundsModule = exports.defaultModule = void 0;
// ------------------ Base ------------------
__exportStar(require("./base/actions/action"), exports);
__exportStar(require("./base/actions/action-dispatcher"), exports);
__exportStar(require("./base/actions/action-handler"), exports);
__exportStar(require("./base/actions/diagram-locker"), exports);
__exportStar(require("./base/animations/animation-frame-syncer"), exports);
__exportStar(require("./base/animations/animation"), exports);
__exportStar(require("./base/animations/easing"), exports);
__exportStar(require("./base/commands/command"), exports);
__exportStar(require("./base/commands/command-registration"), exports);
__exportStar(require("./base/commands/command-stack-options"), exports);
__exportStar(require("./base/commands/command-stack"), exports);
__exportStar(require("./base/features/initialize-canvas"), exports);
__exportStar(require("./base/features/set-model"), exports);
__exportStar(require("./base/model/smodel-factory"), exports);
__exportStar(require("./base/model/smodel-utils"), exports);
__exportStar(require("./base/model/smodel"), exports);
__exportStar(require("./base/ui-extensions/ui-extension-registry"), exports);
__exportStar(require("./base/ui-extensions/ui-extension"), exports);
__exportStar(require("./base/views/key-tool"), exports);
__exportStar(require("./base/views/mouse-tool"), exports);
__exportStar(require("./base/views/thunk-view"), exports);
__exportStar(require("./base/views/view"), exports);
__exportStar(require("./base/views/viewer-cache"), exports);
__exportStar(require("./base/views/viewer-options"), exports);
__exportStar(require("./base/views/viewer"), exports);
__exportStar(require("./base/views/vnode-postprocessor"), exports);
__exportStar(require("./base/views/vnode-utils"), exports);
__exportStar(require("./base/types"), exports);
const di_config_1 = __importDefault(require("./base/di.config"));
exports.defaultModule = di_config_1.default;
// ------------------ Features ------------------
__exportStar(require("./features/bounds/bounds-manipulation"), exports);
__exportStar(require("./features/bounds/hidden-bounds-updater"), exports);
__exportStar(require("./features/bounds/layout"), exports);
__exportStar(require("./features/bounds/model"), exports);
__exportStar(require("./features/bounds/vbox-layout"), exports);
__exportStar(require("./features/bounds/hbox-layout"), exports);
__exportStar(require("./features/bounds/stack-layout"), exports);
__exportStar(require("./features/bounds/views"), exports);
__exportStar(require("./features/button/button-handler"), exports);
__exportStar(require("./features/button/model"), exports);
__exportStar(require("./features/command-palette/action-providers"), exports);
__exportStar(require("./features/command-palette/command-palette"), exports);
__exportStar(require("./features/context-menu/context-menu-service"), exports);
__exportStar(require("./features/context-menu/menu-providers"), exports);
__exportStar(require("./features/context-menu/mouse-listener"), exports);
__exportStar(require("./features/edge-layout/di.config"), exports);
__exportStar(require("./features/edge-layout/edge-layout"), exports);
__exportStar(require("./features/edge-layout/model"), exports);
__exportStar(require("./features/edit/create"), exports);
__exportStar(require("./features/edit/create-on-drag"), exports);
__exportStar(require("./features/edit/di.config"), exports);
__exportStar(require("./features/edit/delete"), exports);
__exportStar(require("./features/edit/edit-label"), exports);
__exportStar(require("./features/edit/edit-label-ui"), exports);
__exportStar(require("./features/edit/edit-routing"), exports);
__exportStar(require("./features/edit/model"), exports);
__exportStar(require("./features/edit/reconnect"), exports);
__exportStar(require("./features/expand/expand"), exports);
__exportStar(require("./features/expand/model"), exports);
__exportStar(require("./features/expand/views"), exports);
__exportStar(require("./features/export/export"), exports);
__exportStar(require("./features/export/model"), exports);
__exportStar(require("./features/export/svg-exporter"), exports);
__exportStar(require("./features/export/svg-export-postprocessor"), exports);
__exportStar(require("./features/fade/fade"), exports);
__exportStar(require("./features/fade/model"), exports);
__exportStar(require("./features/hover/hover"), exports);
__exportStar(require("./features/hover/model"), exports);
__exportStar(require("./features/decoration/model"), exports);
__exportStar(require("./features/decoration/views"), exports);
__exportStar(require("./features/decoration/decoration-placer"), exports);
__exportStar(require("./features/edge-intersection/intersection-finder"), exports);
__exportStar(require("./features/edge-intersection/sweepline"), exports);
__exportStar(require("./features/edge-junction/junction-finder"), exports);
__exportStar(require("./features/edge-junction/junction-postprocessor"), exports);
__exportStar(require("./features/move/model"), exports);
__exportStar(require("./features/move/move"), exports);
__exportStar(require("./features/move/snap"), exports);
__exportStar(require("./features/nameable/model"), exports);
__exportStar(require("./features/open/open"), exports);
__exportStar(require("./features/open/model"), exports);
__exportStar(require("./features/projection/model"), exports);
__exportStar(require("./features/projection/views"), exports);
__exportStar(require("./features/routing/anchor"), exports);
__exportStar(require("./features/routing/abstract-edge-router"), exports);
__exportStar(require("./features/routing/bezier-anchors"), exports);
__exportStar(require("./features/routing/bezier-edge-router"), exports);
__exportStar(require("./features/routing/manhattan-anchors"), exports);
__exportStar(require("./features/routing/manhattan-edge-router"), exports);
__exportStar(require("./features/routing/model"), exports);
__exportStar(require("./features/routing/polyline-anchors"), exports);
__exportStar(require("./features/routing/polyline-edge-router"), exports);
__exportStar(require("./features/routing/routing"), exports);
__exportStar(require("./features/routing/views"), exports);
__exportStar(require("./features/select/model"), exports);
__exportStar(require("./features/select/select"), exports);
__exportStar(require("./features/undo-redo/undo-redo"), exports);
__exportStar(require("./features/update/model-matching"), exports);
__exportStar(require("./features/update/update-model"), exports);
__exportStar(require("./features/viewport/center-fit"), exports);
__exportStar(require("./features/viewport/model"), exports);
__exportStar(require("./features/viewport/scroll"), exports);
__exportStar(require("./features/viewport/viewport-root"), exports);
__exportStar(require("./features/viewport/viewport"), exports);
__exportStar(require("./features/viewport/zoom"), exports);
__exportStar(require("./features/zorder/zorder"), exports);
const di_config_2 = __importDefault(require("./features/bounds/di.config"));
exports.boundsModule = di_config_2.default;
const di_config_3 = __importDefault(require("./features/button/di.config"));
exports.buttonModule = di_config_3.default;
const di_config_4 = __importDefault(require("./features/command-palette/di.config"));
exports.commandPaletteModule = di_config_4.default;
const di_config_5 = __importDefault(require("./features/context-menu/di.config"));
exports.contextMenuModule = di_config_5.default;
const di_config_6 = __importDefault(require("./features/decoration/di.config"));
exports.decorationModule = di_config_6.default;
const di_config_7 = __importDefault(require("./features/edge-intersection/di.config"));
exports.edgeIntersectionModule = di_config_7.default;
const di_config_8 = __importDefault(require("./features/edge-junction/di.config"));
exports.edgeJunctionModule = di_config_8.default;
const di_config_9 = __importDefault(require("./features/edge-layout/di.config"));
exports.edgeLayoutModule = di_config_9.default;
const di_config_10 = __importDefault(require("./features/expand/di.config"));
exports.expandModule = di_config_10.default;
const di_config_11 = __importDefault(require("./features/export/di.config"));
exports.exportModule = di_config_11.default;
const di_config_12 = __importDefault(require("./features/fade/di.config"));
exports.fadeModule = di_config_12.default;
const di_config_13 = __importDefault(require("./features/hover/di.config"));
exports.hoverModule = di_config_13.default;
const di_config_14 = __importDefault(require("./features/move/di.config"));
exports.moveModule = di_config_14.default;
const di_config_15 = __importDefault(require("./features/open/di.config"));
exports.openModule = di_config_15.default;
const di_config_16 = __importDefault(require("./features/routing/di.config"));
exports.routingModule = di_config_16.default;
const di_config_17 = __importDefault(require("./features/select/di.config"));
exports.selectModule = di_config_17.default;
const di_config_18 = __importDefault(require("./features/undo-redo/di.config"));
exports.undoRedoModule = di_config_18.default;
const di_config_19 = __importDefault(require("./features/update/di.config"));
exports.updateModule = di_config_19.default;
const di_config_20 = __importDefault(require("./features/viewport/di.config"));
exports.viewportModule = di_config_20.default;
const di_config_21 = __importDefault(require("./features/zorder/di.config"));
exports.zorderModule = di_config_21.default;
// ------------------ Graph ------------------
__exportStar(require("./graph/sgraph"), exports);
__exportStar(require("./graph/views"), exports);
// ------------------ Library ------------------
__exportStar(require("./lib/modules"), exports);
__exportStar(require("./lib/generic-views"), exports);
__exportStar(require("./lib/html-views"), exports);
__exportStar(require("./lib/jsx"), exports);
__exportStar(require("./lib/model"), exports);
__exportStar(require("./lib/svg-views"), exports);
// ------------------ Model Source ------------------
__exportStar(require("./model-source/commit-model"), exports);
__exportStar(require("./model-source/diagram-server"), exports);
__exportStar(require("./model-source/local-model-source"), exports);
__exportStar(require("./model-source/logging"), exports);
__exportStar(require("./model-source/model-source"), exports);
__exportStar(require("./model-source/websocket"), exports);
const di_config_22 = __importDefault(require("./model-source/di.config"));
exports.modelSourceModule = di_config_22.default;
// ------------------ Utilities ------------------
__exportStar(require("./utils/browser"), exports);
__exportStar(require("./utils/codicon"), exports);
__exportStar(require("./utils/color"), exports);
__exportStar(require("./utils/geometry"), exports);
__exportStar(require("./utils/inversify"), exports);
__exportStar(require("./utils/logging"), exports);
__exportStar(require("./utils/registry"), exports);
