// sprotty/lib/features/decoration/model.js
"use strict";
/********************************************************************************
 * Copyright (c) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.SIssue = exports.SIssueMarker = exports.SIssueMarkerImpl = exports.SDecoration = exports.isDecoration = exports.decorationFeature = void 0;
const model_1 = require("../bounds/model");
const model_2 = require("../hover/model");
exports.decorationFeature = Symbol('decorationFeature');
function isDecoration(e) {
    return e.hasFeature(exports.decorationFeature);
}
exports.isDecoration = isDecoration;
class SDecoration extends model_1.SShapeElementImpl {
}
exports.SDecoration = SDecoration;
SDecoration.DEFAULT_FEATURES = [exports.decorationFeature, model_1.boundsFeature, model_2.hoverFeedbackFeature, model_2.popupFeature];
class SIssueMarkerImpl extends SDecoration {
}
exports.SIssueMarkerImpl = SIssueMarkerImpl;
/** @deprecated Use SIssueMarkerImpl instead. */
exports.SIssueMarker = SIssueMarkerImpl;
/**
 * @deprecated Use the definition from `sprotty-protocol` instead.
 */
class SIssue {
}
exports.SIssue = SIssue;
