// sprotty/lib/features/zorder/zorder.js
"use strict";
/********************************************************************************
 * Copyright (c) 2019 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BringToFrontCommand = void 0;
const inversify_1 = require("inversify");
const actions_1 = require("sprotty-protocol/lib/actions");
const types_1 = require("../../base/types");
const smodel_1 = require("../../base/model/smodel");
const command_1 = require("../../base/commands/command");
const model_1 = require("../routing/model");
let BringToFrontCommand = class BringToFrontCommand extends command_1.Command {
    constructor(action) {
        super();
        this.action = action;
        this.selected = [];
    }
    execute(context) {
        const model = context.root;
        this.action.elementIDs.forEach(id => {
            const element = model.index.getById(id);
            if (element instanceof model_1.SRoutableElementImpl) {
                if (element.source)
                    this.addToSelection(element.source);
                if (element.target)
                    this.addToSelection(element.target);
            }
            if (element instanceof smodel_1.SChildElementImpl) {
                this.addToSelection(element);
            }
            this.includeConnectedEdges(element);
        });
        return this.redo(context);
    }
    includeConnectedEdges(element) {
        if (element instanceof model_1.SConnectableElementImpl) {
            element.incomingEdges.forEach(edge => this.addToSelection(edge));
            element.outgoingEdges.forEach(edge => this.addToSelection(edge));
        }
        if (element instanceof smodel_1.SParentElementImpl) {
            for (const child of element.children) {
                this.includeConnectedEdges(child);
            }
        }
    }
    addToSelection(element) {
        this.selected.push({
            element: element,
            index: element.parent.children.indexOf(element)
        });
    }
    undo(context) {
        for (let i = this.selected.length - 1; i >= 0; i--) {
            const selection = this.selected[i];
            const element = selection.element;
            element.parent.move(element, selection.index);
        }
        return context.root;
    }
    redo(context) {
        for (let i = 0; i < this.selected.length; i++) {
            this.bringToFront(this.selected[i]);
        }
        return context.root;
    }
    bringToFront(selection) {
        const element = selection.element;
        const childrenLength = element.parent.children.length;
        element.parent.move(element, childrenLength - 1);
    }
};
exports.BringToFrontCommand = BringToFrontCommand;
BringToFrontCommand.KIND = actions_1.BringToFrontAction.KIND;
exports.BringToFrontCommand = BringToFrontCommand = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(types_1.TYPES.Action)),
    __metadata("design:paramtypes", [Object])
], BringToFrontCommand);
